plugin/acl/acl.go

```diff
@@ -0,0 +1,99 @@
+package acl
+
+import (
+	"context"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/metrics"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/infobloxopen/go-trees/iptree"
+	"github.com/miekg/dns"
+)
+
+var log = clog.NewWithPlugin("acl")
+
+type acl struct {
+	Next plugin.Handler
+
+	Rules []Rule
+}
+
+// Rule defines a list of Zones and some ACL policies which will be
+// enforced on them.
+type Rule struct {
+	Zones    []string
+	Policies []Policy
+}
+
+// Policy defines the ACL policy for DNS queries.
+// A policy performs the specified action (block/allow) on all DNS queries
+// matched by source IP or QTYPE.
+type Policy struct {
+	action int
+	qtypes map[uint16]struct{}
+	filter *iptree.Tree
+}
+
+const (
+	// Allow allows authorized queries to recurse.
+	Allow = iota
+	// Block blocks unauthorized queries towards protected DNS zones.
+	Block
+)
+
+func (a acl) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	for _, rule := range a.Rules {
+		// check zone.
+		zone := plugin.Zones(rule.Zones).Matches(state.Name())
+		if zone == "" {
+			continue
+		}
+
+		if shouldBlock(rule.Policies, w, r) {
+			m := new(dns.Msg)
+			m.SetRcode(r, dns.RcodeRefused)
+			w.WriteMsg(m)
+			RequestBlockCount.WithLabelValues(metrics.WithServer(ctx), zone).Inc()
+			// TODO: should we return Success here? (@ihac)
+			return dns.RcodeSuccess, nil
```

Comment:
@miekg Does it make sense to return success here? WDYT.

Reply:
you return success to indicate you wrote to the client. See plugin.ClientWrite

Reply:
`return 0, nil` can be used here, and the comment removed

Reply:
So keep `RcodeSuccess` here?